(PROFILES,) = glob_wildcards(config["bracken_dir"] + "/{p}.txt")


rule all:
    input:
        expand("{dir}/{id}_genomes.fasta", dir=config["profiles_dir"], id=PROFILES),

# input:
#     expand(
#         "{s}/{a}_{m}_reads_R1.fastq",
#         s=config["simulated_dir"],
#         m=config["model"],
#         a=config["abundances"],
#     ),
#     expand(
#         "{s}/{a}_{m}_reads_R2.fastq",
#         s=config["simulated_dir"],
#         m=config["model"],
#         a=config["abundances"],
#     ),


rule make_profiles:
    input:
        "{dir}/{id}.txt",
    output:
        "{dir}/{id}_profile.txt",
        "{dir}/{id}_files.txt",
    params:
        bracken_profiler=config["bracken_profiler"],
        refseq=config["refseq_info"],
        out_dir=config["profiles_dir"],
    shell:
        """
        {params.bracken_profiler} \
            -t {params.refseq} \
            -o {params.out_dir} \
            {input}
        """


rule cat_genomes:
    input:
        "{dir}/{id}_files.txt",
    output:
        "{dir}/{id}_genomes.fasta",
    params:
        cat_genomes=config["cat_genomes"],
        out_dir=config["profiles_dir"],
        parent=config["refseq_dir"]
    shell:
        """
        {params.cat_genomes} \
            -p {params.parent} \
            -o {params.out_dir} \
            {input}
        """

# rule simulate_reads:
#     input:
#         genomes=config["genomes_dir"] + "/cat_genomes.fasta",
#         abundances=config["profiles_dir"]+"/profile_{abundance}.txt",
#     output:
#         config["simulated_dir"] + "/{abundance}_{model}_reads_R1.fastq",
#         config["simulated_dir"] + "/{abundance}_{model}_reads_R2.fastq",
#     params:
#         simulator=config["simulator"],
#         simulated_dir=config["simulated_dir"],
#         cpus=config["simulator_cpus"],
#         reads=config["n_reads"],
#     shell:
#         """
#         source ~/.bashrc
#         source activate insilicoseq_env
#         {params.simulator} {params.cpus} {params.reads} \
#         --model {wildcards.model} \
#         --abundance_file {input.abundances} \
#         --output {params.simulated_dir}/{wildcards.abundance}_{wildcards.model}_reads \
#          --genomes {input.genomes}
#         """
