(IDS,) = glob_wildcards(config["genomes_dir"] + "/{id}.fna")


rule all:
    input:
        expand(
            "{s}/{a}_{m}_reads_R1.fastq",
            s=config["simulated_dir"],
            m=config["model"],
            a=config["abundances"],
        ),
        expand(
            "{s}/{a}_{m}_reads_R2.fastq",
            s=config["simulated_dir"],
            m=config["model"],
            a=config["abundances"],
        ),


rule cat_genomes:
    input:
        expand("{dir}/{id}.fna", dir=config["genomes_dir"], id=IDS),
    output:
        config["genomes_dir"] + "/cat_genomes.fasta",
    params:
        genomes_dir=config["genomes_dir"],
    shell:
        """
        cd {params.genomes_dir}
        ./cat_genomes.sh
        """


rule simulate_reads:
    input:
        genomes=config["genomes_dir"] + "/cat_genomes.fasta",
        abundances="profiles/profile_{abundance}.txt",
    output:
        config["simulated_dir"] + "/{abundance}_{model}_reads_R1.fastq",
        config["simulated_dir"] + "/{abundance}_{model}_reads_R2.fastq",
    params:
        simulator=config["simulator"],
        simulated_dir=config["simulated_dir"],
        cpus=config["simulator_cpus"],
        reads=config["n_reads"],
    shell:
        """
        source ~/.bashrc
        source activate insilicoseq_env

        {params.simulator} {params.cpus} {params.reads} --model {wildcards.model} --abundance_file {input.abundances} --output {params.simulated_dir}/{wildcards.abundance}_{wildcards.model}_reads --genomes {input.genomes}
        """
