---
title: "CRC Classification"
output: html_notebook
---

# {.tabset}

## Overview

This notebook is for processing the classification of the data in the CRC set.


```{r, include = FALSE}
library(dplyr)
library(ggplot2)
library(magrittr)
library(readr)
library(UpSetR)
library(stringr)
library(tidyr)
library(tidytable)

# Set ggplot2 theme for whole R session
theme_set(theme_light() +
            theme(plot.title = element_text(hjust = 0.5),
                  plot.subtitle = element_text(hjust = 0.5)))
```

```{r, include=FALSE}
reference_free <- c("DeepVirFinder", "Seeker", "VirFinder")
reference_based <-
  c("MARVEL",
    "MetaPhinder",
    "VIBRANT",
    "viralVerify",
    "VirSorter",
    "VirSorter2")

raw_crc_df <-
  read_csv("../../data/classified_crc/combined_out/combined.csv",
           show_col_types = FALSE)
raw_lengths <-
  read_csv("../../data/crc_data/contig_counts/contig_lengths.csv",
           show_col_types = FALSE)

lengths <- raw_lengths %>%
  mutate(
    contig = str_extract(contig, "k[\\d]+_[\\d]+"),
    sample_contig = paste(sameple, contig, sep = "_")
  ) %>%
  select(-sameple, -contig)

rm(raw_lengths)

crc_df <- raw_crc_df %>%
  rename(
    "DeepVirFinder" = "dvf",
    "MARVEL" = "marvel",
    "MetaPhinder" = "metaphinder",
    "Seeker" = "seeker",
    "VIBRANT" = "vibrant",
    "viralVerify" = "viralverify",
    "VirFinder" = "virfinder",
    "VirSorter" = "virsorter",
    "VirSorter2" = "virsorter2"
  ) %>%
  mutate_all(~ replace(., is.na(.), "non-viral")) %>%
  rename(sample = metagenome, contig = record) %>%
  mutate(sample_contig = paste(sample, contig, sep = "_")) %>%
  select(-sample, -contig) %>%
  mutate_all(~ replace(., .  == "viral", 1)) %>%
  mutate_all(~ replace(., .  == "non-viral", 0)) %>%
  right_join(lengths)

rm(raw_crc_df, lengths)

crc_matrix_df <- crc_df %>%
  select(-sample_contig,-length) %>%
  mutate_if(is.character, as.numeric) %>%
  as.data.frame(crc_matrix)

write_csv(crc_df, "data/processed_df.csv")

short_contigs <- crc_df %>%
  filter(length < 5000) %>%
  select(-sample_contig,-length) %>%
  mutate_if(is.character, as.numeric) %>%
  as.data.frame(crc_matrix)

write_csv(short_contigs, file = "data/short_contigs.csv")
rm(short_contigs)

long_contigs <- crc_df %>%
  filter(length >= 5000) %>%
  select(-sample_contig,-length) %>%
  mutate_if(is.character, as.numeric) %>%
  as.data.frame(crc_matrix)

write_csv(long_contigs, file = "data/long_contigs.csv")
rm(long_contigs)
rm(crc_df)
```

## Upset Plots

### All Tools

First, plotting for all tools.

The total number of possible intersections for 9 tools is 511, so not all intersections are shown.

Here, the number of intersections is set to 51 (The top 10% of largest intersections).

```{r, echo = FALSE}
upset(
  crc_matrix_df,
  order.by = "freq",
  empty.intersections = TRUE,
  sets = colnames(crc_matrix_df),
  show.numbers = FALSE,
  nintersects = 51
)
```

### Reference-free Tools

Since the reference-free tools identify so many more phages, I decided to create separate plots for reference-based and reference-free tools.


```{r, echo = FALSE}
upset(
  crc_matrix_df %>%  select(all_of(reference_free)),
  order.by = "freq",
  empty.intersections = TRUE,
  show.numbers = FALSE
)
```

### Reference-based Tools

```{r, echo = FALSE}
upset(
  crc_matrix_df %>%  select(all_of(reference_based)),
  order.by = "freq",
  empty.intersections = TRUE,
  show.numbers = FALSE,
  nintersects = 25
)
```

Here is the same plot, but excluding MetaPhinder since it identifies the most phages.

```{r, echo = FALSE}
upset(
  crc_matrix_df %>% select(all_of(reference_based), -MetaPhinder),
  order.by = "freq",
  empty.intersections = TRUE,
  show.numbers = FALSE,
  nintersects = 25
)

```
### Short vs Long

#### Contigs >= 5000 nt


```{r, include=FALSE}
rm(crc_matrix_df)

long_contigs <-
  read_csv("data/long_contigs.csv", show_col_types = FALSE) %>%
  as.data.frame()
```

```{r, echo=FALSE}
upset(
  long_contigs,
  order.by = "freq",
  empty.intersections = TRUE,
  sets = colnames(long_contigs),
  show.numbers = FALSE,
  nintersects = 51
)
```

```{r, include=FALSE}
rm(long_contigs)

short_contigs <-
  read_csv("data/short_contigs.csv", show_col_types = FALSE) %>%
  as.data.frame()
```

#### Contigs < 5000 nt

```{r, echo=FALSE}
upset(
  short_contigs,
  order.by = "freq",
  empty.intersections = TRUE,
  sets = colnames(short_contigs),
  show.numbers = FALSE,
  nintersects = 51
)
```

## Length Correlations

```{r, include=FALSE}
bin_lengths <- function(df, step, min, max) {
  df <- df %>%
    mutate(log_len = log(length, base = 10))
  
  df$length_bin = min
  
  for (bin in seq(min, max, step)) {
    df$length_bin[df$log_len >= bin] = bin
  }
  
  df %>% 
    select(-log_len, -length)
}
```


```{r, include=FALSE}
rm(short_contigs)

crc_df <- read_csv("data/processed_df.csv", show_col_types = FALSE)
crc_df <- crc_df %>% 
  separate(sample_contig, c("sample", "contig"), sep = 15) %>% 
  bin_lengths(0.5, 2.5, 5.0) %>% 
  group_by(sample, length_bin) %>%
  summarize(
    total_count = n(),
    dvf_sum = sum(DeepVirFinder),
    meta_sum = sum(MetaPhinder),
    seeker_sum = sum(Seeker),
    vibrant_sum = sum(VIBRANT),
    viralverify_sum = sum(viralVerify),
    virfinder_sum = sum(VirFinder),
    virsorter_sum = sum(VirSorter),
    virsorter2_sum = sum(VirSorter2),
    marvel_sum = sum(MARVEL)
  ) %>%
  pivot_longer(cols = ends_with("sum"),
               names_to = "tool",
               values_to = "predicted") %>%
  mutate(
    tool = case_when(
      tool == "dvf_sum" ~ "DeepVirFinder",
      tool == "meta_sum" ~ "MetaPhinder",
      tool == "seeker_sum" ~ "Seeker",
      tool == "vibrant_sum" ~ "VIBRANT",
      tool == "viralverify_sum" ~ "viralVerify",
      tool == "virfinder_sum" ~ "VirFinder",
      tool == "virsorter_sum" ~ "VirSorter",
      tool == "virsorter2_sum" ~ "VirSorter2",
      tool == "marvel_sum" ~ "MARVEL"
    ),
    tool = factor(
      tool,
      levels = c(
        "VirFinder",
        "DeepVirFinder",
        "Seeker",
        "VirSorter",
        "MetaPhinder",
        "MARVEL",
        "viralVerify",
        "VIBRANT",
        "VirSorter2"
      )
    ),
    predicted_prop = predicted / total_count
  )

```

```{r, echo=FALSE}
crc_df %>%
  ggplot(aes(x = factor(length_bin), y = predicted_prop)) +
  facet_wrap(~ tool) +
  geom_boxplot(
    color = "grey30",
    lwd = 0.25,
    outlier.alpha = 0.7,
    outlier.size = 0.85,
    outlier.color = "gray40"
  ) +
  labs(x = "log(Contig Length) Bin",
       y = "Phage Proportion of Total Contigs")

ggsave("figures/prop_phage_length.png", height=6, width=6)
ggsave("figures/prop_phage_length.svg", height=6, width=6)
```

```{r, echo=FALSE}
crc_df %>%
  ggplot(aes(x = factor(length_bin), y = predicted)) +
  facet_wrap(~ tool) +
  geom_boxplot(
    color = "grey30",
    lwd = 0.25,
    outlier.alpha = 0.7,
    outlier.size = 0.85,
    outlier.color = "gray40"
  ) +
  labs(x = "log(Contig Length) Bin",
       y = "Number of Predicted Phage Contigs")

ggsave("figures/phage_length.png",
       height = 6,
       width = 6)
ggsave("figures/phage_length.svg",
       height = 6,
       width = 6)
```

```{r,echo=FALSE}
crc_df %>%
  ggplot(aes(x = factor(length_bin), y = predicted)) +
  facet_wrap(~ tool) +
  geom_boxplot(
    color = "grey30",
    lwd = 0.25,
    outlier.alpha = 0.7,
    outlier.size = 0.85,
    outlier.color = "gray40"
  ) +
  scale_y_continuous(trans = "log10") +
  labs(x = "log(Contig Length) Bin",
       y = "Number of Predicted Phage Contigs")

ggsave("figures/log_phage_length.png",
       height = 6,
       width = 6)
ggsave("figures/log_phage_length.svg",
       height = 6,
       width = 6)
```

